[tagger] Enable remote tagger by default

juliogreff opened the pull request:
### What does this PR do?

This re-applies changes from #9313 that were subsequently reverted due to a memory leak (fixed in #9452). It also adds a fallback in the trace-agent for backwards compatibility, in case the remote tagger is not available if running standalone and remote tagger is not disabled.

### Describe how to test/QA your changes

* Test that the remote tagger is actually enabled in process, trace, and security agent.
* Check that tagging still works for those agents. This is especially important for environments in which the remote tagger is new: ECS EC2 and ECS Fargate.
* Test that the trace-agent falls back to the local tagger if it's running standalone.

juliogreff committed:
[tagger] Enable remote tagger by default (#9313)

juliogreff committed:
[tagger] Fall back to local tagger in trace agent

If the trace agent cannot connect to the remote tagger, for instance if
it is running stand alone, we now fall back to the local tagger. This is
only necessary as the remote tagger is now the default as of previous
commit.

gbbr commented on pkg/trace/agent/run.go:
```diff
@@ -144,17 +144,29 @@ func Run(ctx context.Context) {
+	remoteTagger := coreconfig.Datadog.GetBool("apm_config.remote_tagger")
+	if remoteTagger {
+		tagger.SetDefaultTagger(remote.NewTagger())
+		err = tagger.Init()
+		if err != nil {
```
```suggestion
		if err := tagger.Init(); err != nil {
```

gbbr commented on pkg/trace/agent/run.go:
```diff
@@ -144,17 +144,29 @@ func Run(ctx context.Context) {
 
-		t = local.NewTagger(collectors.DefaultCatalog)
+		tagger.SetDefaultTagger(local.NewTagger(collectors.DefaultCatalog))
+		err = tagger.Init()
+		if err != nil {
```
Same here

juliogreff committed:
Apply PR suggestions

djmitche commented on pkg/tagger/global.go:
```diff
@@ -47,7 +47,9 @@ var tlmUDPOriginDetectionError = telemetry.NewCounter("dogstatsd", "udp_origin_d
 // Init must be called once config is available, call it in your cmd
-func Init() {
+func Init() error {
+	var initErr error
```
Should this be a global?  Otherwise calls to Init after the first (of which there must be some, or `sync.Once` wouldn't be necessary) will return no error when in fact an error occurred.

juliogreff committed:
Make initErr global

This remembers the error between different calls to Init, instead of
pretending that follow up calls were successful.

clamoriniere commented on cmd/agent/common/loader.go:
```diff
@@ -30,7 +31,10 @@ func LoadComponents(confdPath string) {
 	// be the first subscribed to metadata store to avoid race conditions.
 	tagger.SetDefaultTagger(local.NewTagger(collectors.DefaultCatalog))
-	tagger.Init()
+	err := tagger.Init()
+	if err != nil {
```
small suggestion
```suggestion
	if err := tagger.Init(); err != nil {
```

clamoriniere commented on cmd/dogstatsd/main.go:
```diff
@@ -213,7 +213,10 @@ func runAgent(ctx context.Context) (err error) {
 
 		tagger.SetDefaultTagger(local.NewTagger(collectors.DefaultCatalog))
-		tagger.Init()
+		err := tagger.Init()
+		if err != nil {
```
```suggestion
		if err := tagger.Init(); err != nil {
```

clamoriniere reviewed (approved):
small coding style nits.

juliogreff committed:
Apply suggestions from code review

Co-authored-by: Cedric Lamoriniere <cedric.lamoriniere@datadoghq.com>
